staging/src/k8s.io/apiserver/pkg/endpoints/discovery/version.go

```diff
@@ -78,6 +80,57 @@ func (s *APIVersionHandler) handle(req *restful.Request, resp *restful.Response)
 }
 
 func (s *APIVersionHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
-	responsewriters.WriteObjectNegotiated(s.serializer, negotiation.DefaultEndpointRestrictions, schema.GroupVersion{}, w, req, http.StatusOK,
-		&metav1.APIResourceList{GroupVersion: s.groupVersion.String(), APIResources: s.apiResourceLister.ListAPIResources()})
+	reqURL := req.URL
+	if reqURL == nil {
+		// Can not find documentation guaranteeing the non-nility of reqURL.
+		w.WriteHeader(500)
+		return
+	}
+
+	// Get current resource list and hash
+	resourceList, hash := s.apiResourceLister.ListAPIResources()
+
+	if providedHash := reqURL.Query().Get("hash"); len(providedHash) > 0 && len(hash) > 0 {
+		if hash == providedHash {
+			// ETag must be enclosed in double quotes:
+			// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag
+			w.Header().Set("Etag", strconv.Quote(hash))
```

Comment:
Should we set the ETag regardless of whether a hash was provided?

Reply:
Looks like that is the behavior of openapi (kube-openapi sets E-tag unconditionally)

https://github.com/kubernetes/kube-openapi/blob/master/pkg/handler3/handler.go#L233

I am not 100% sure on the correct behavior. Seem like ETag by itself with no Cache-Control headers should be fine?

Reply:
Yeah, etag indicates the version of the returned document, I think we can safely return it even without the query parameter (in which case we should also not have the cache-control set to be immutable).

Reply:
On the other hand you're doing a redirect if the url doesn't have the hash so I guess it's irrelevant.